src/domain/tradingview/TVDataProvider.ts

```diff
@@ -0,0 +1,183 @@
+  period: "",
+  data: [],
+  ticker: "",
+};
+
+export class TVDataProvider {
+  lastBar: Bar | null;
+  startTime: number;
+  lastTicker: string;
+  lastPeriod: string;
+  barsInfo: {
+    period: string;
+    data: Bar[];
+    ticker: string;
+  };
+
+  constructor() {
+    this.lastBar = null;
+    this.startTime = 0;
+    this.lastTicker = "";
+    this.lastPeriod = "";
+    this.barsInfo = initialHistoryBarsInfo;
+  }
+
+  async getCurrentPriceOfToken(chainId: number, ticker: string): Promise<BigNumberish> {
+    return getCurrentPriceOfToken(chainId, ticker);
+  }
+
+  async getTokenLastBars(chainId: number, ticker: string, period: string, limit: number): Promise<Bar[]> {
+    return getLimitChartPricesFromStats(chainId, ticker, period, limit);
+  }
+
+  async getTokenHistoryBars(
+    chainId: number,
+    ticker: string,
+    period: string,
+    periodParams: PeriodParams,
+    shouldRefetchBars: { current: boolean }
+  ): Promise<Bar[]> {
+    const barsInfo = this.barsInfo;
+    if (
+      !barsInfo.data.length ||
+      barsInfo.ticker !== ticker ||
+      barsInfo.period !== period ||
+      shouldRefetchBars.current
+    ) {
+      try {
+        const bars = await getTokenChartPrice(chainId, ticker, period);
+        const filledBars = fillBarGaps(bars, CHART_PERIODS[period]);
+        const currentCandleTime = getCurrentCandleTime(period);
+        const lastCandleTime = currentCandleTime - CHART_PERIODS[period];
+        const lastBar = filledBars[filledBars.length - 1];
+        if (lastBar.time === currentCandleTime || lastBar.time === lastCandleTime) {
+          this.lastBar = { ...lastBar, ticker };
+        }
+        this.barsInfo.data = filledBars;
+        this.barsInfo.ticker = ticker;
+        this.barsInfo.period = period;
+        shouldRefetchBars.current = false;
+      } catch (error) {
+        // eslint-disable-next-line no-console
+        console.error(error);
+        this.barsInfo = initialHistoryBarsInfo;
+      }
+    }
+
+    const { from, to, countBack } = periodParams;
+    const toWithOffset = to + timezoneOffset;
+    const fromWithOffset = from + timezoneOffset;
+    const bars = barsInfo.data.filter((bar) => bar.time > fromWithOffset && bar.time <= toWithOffset);
+
+    // if no bars returned, return empty array
+    if (!bars.length) {
+      return [];
+    }
+
+    // if bars are fewer than countBack, return all of them
+    if (bars.length < countBack) {
+      return bars;
+    }
+
+    // if bars are more than countBack, return latest bars
+    return bars.slice(bars.length - countBack, bars.length);
+  }
+
+  async getBars(
+    chainId: number,
+    ticker: string,
+    resolution: string,
+    isStable: boolean,
+    periodParams: PeriodParams,
+    shouldRefetchBars: { current: boolean }
+  ) {
+    const period = SUPPORTED_RESOLUTIONS[resolution];
+    const { countBack } = periodParams;
+
+    try {
+      const bars = isStable
+        ? getStablePriceData(period, countBack)
```

Comment:
not a priority, but G wants to get stable tokens prices for synthetics for from ChainLink instead of hardoced values, so in the future i will probably try to to make getTokenHistoryBars unified for stable and not stable tokens